backend/program/__init__.py

```diff
@@ -24,47 +46,209 @@ def __init__(self, args):
+                max_instances=1,
+                replace_existing=True,  # Replace existing jobs with the same ID
+                next_run_time=datetime.now() if service_cls != PlexLibrary else None
+            )
+            logger.info("Scheduled %s to run every %s seconds.", service_cls.__name__, update_interval)
+        return None
+
+    def _process_future_item(self, future: Future, service: Service, input_item: MediaItem) -> None:
+        """Callback to add the results from a future emitted by a service to the event queue."""
+        try:
+            for item in future.result():
+                if not isinstance(item, MediaItem):
+                    logger.error("Service %s emitted item %s of type %s, skipping", service.__name__, item, item.__class__.__name__)
+                    continue
+                self.job_queue.put(Event(emitted_by=service, item=item))
+        except Exception as e:
+            logger.error("Service %s failed with exception %s", service.__name__, traceback.format_exc())
+
+    def _service_run_item(self, func: Callable, item: MediaItem | None) -> MediaItemGenerator:
+        """Some services don't intake any items so we want to make sure we don't provide them when the input is None."""
+        if item is None:
+            yield from func()
+        else: 
+            yield from func(item)
+ 
+    def _submit_job(self, service: Service, item: MediaItem | None) -> None:
+        logger.debug(
+            f"Submitting service {service.__name__} to the pool" +
+            (f" with {item.title or item.item_id}" if item else "")
+        )
+        func = self.services[service].run
+        future = self.executor.submit(self._service_run_item, func, item)
+        future.add_done_callback(lambda f: self._process_future_item(f, service, item))
+    
+
     def run(self):
         while self.running:
-            if self.validate():
-                with concurrent.futures.ThreadPoolExecutor(
-                    max_workers=10, thread_name_prefix="Worker"
-                ) as executor:
-                    for item in self.media_items:
-                        executor.submit(item.perform_action, self.core_manager.services)
-            time.sleep(1)
+            if not self.validate():
+                time.sleep(1)
+                continue
+            try:
+                event: Event = self.job_queue.get(timeout=1)
+            except Empty:
+                # Unblock after waiting in case we are no longer supposed to be running
+                continue
+            service, item = event.emitted_by, event.item
+            existing_item = self.media_items.get(item.item_id, None)
+            # we always want to get metadata for content items before we compare to the container. 
+            # we can't just check if the show exists we have to check if it's complete
+            if service in get_args(Content):
+                if not getattr(item, 'title', None):

```

Comment:
im not sure what i think about this, should metadata have an attribute or something? We might need to update existing items that are not fully released etc.